src/c++/perf_analyzer/command_line_parser.cc

```diff
@@ -0,0 +1,1407 @@
+// Copyright 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions
+// are met:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of NVIDIA CORPORATION nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+// OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "command_line_parser.h"
+
+#include <getopt.h>
+
+#include <iomanip>
+#include <iostream>
+#include <string>
+
+#include "perf_analyzer_exception.h"
```

Comment:
I like the idea of using exceptions but I think it should be a separate ticket. We should be either using exceptions in all part of the Perf Analyzer or none. Otherwise, it will be confusing to the developer whether this method returns and error or an exception.

Reply:
You have to start somewhere. Refactoring does not need to be all or nothing. That thought process just forces us to push out changes and tech debt. You need to be able to make small refactors and work towards a goal over time. This is in the complete spirit of this excercise and it handles all cases right now where an error is returned, rather then thrown. 

As long as it is functional and the behavior has not changed, we can make the changes in small iterations over time, instead of all at once.

Reply:
As a dev, when I start working in this codebase after this change is merged, I have to think twice about how to handle the errors when using the PA functions. That is why I think changes like these should be consistent across the component. Especially, since there is not documentation about the functions that return an exception I have to check the function implementation to figure out the error handling method used.

Reply:
No you don't. It is designed to be modular. If you are working in a class, Parser for example, that is not using exceptions, then just return the error. As a dev of Parser, I don't have an idea that the class that managing parser then takes the error and throws an exception with it. That level has been abstracted. You work and develop to your code. 

The way I have done it starts at the top, creating a methodology that has been encapsulated from the child dependencies. The dependencies don't need to know or care how they returning error is handled. The is the responsibility of the parent object calling them.

However, you would be correct if I mixed error handling in single class. It would be confusing if I adding some exceptions for errors in Parser, but not in all cases. That is the worse of both worlds because now I have to both handle the exception and error return.

But the way it is written now, it can be rolled out in stages. For example, I can decide later to add exceptions to Parser. I can remove all error returns from within Parser and the modify to remove error handling in the caller (i.e. PerfAnalyzer). 

In fact, in this manner, I can roll out Exception handling in stages, which simplifies the story and code review. I can add exceptions to Parser in one PR, to ClientBackend in another, and to InferenceProfiler in another. Doing this separately limits the scope of the change, makes the review easier.

For now, it is not difficult to understand. If you need to return an error in PerfAnalyzer or CLParser, then throw an exception. If in any other class, return the error. Do what that class is doing.

Reply:
Ok, sounds good. Resolving this thread.